src/traits.rs

```diff
@@ -0,0 +1,35 @@
+use crate::{evaluation, providers::traits::FeatureProvider, ClientMetaData, EvaluationDetails};
+
+pub trait ClientTraits<C>
+where
+    C: FeatureProvider,
+{
+    fn new(name: String, provider: C) -> Self;
+    fn meta_data(&self) -> ClientMetaData;
+    fn set_evaluation_context(&mut self, eval_ctx: evaluation::EvaluationContext);
+    fn evaluation_context(&self) -> evaluation::EvaluationContext;
+    fn evaluate<T>(
+        &self,
+        flag: String,
+        default_value: T,
+        eval_ctx: evaluation::EvaluationContext,
+    ) -> (EvaluationDetails<T>, anyhow::Error)
+    where
+        T: Clone;
+    fn value<T>(
+        &self,
+        flag: String,
+        default_value: T,
+        eval_ctx: evaluation::EvaluationContext,
+    ) -> (T, anyhow::Error)
+    where
+        T: Clone;
+    fn value_details<T>(
+        &self,
+        flag: String,
+        default_value: T,
+        eval_ctx: evaluation::EvaluationContext,
+    ) -> (EvaluationDetails<T>, anyhow::Error)
```

Comment:
```suggestion
    ) -> Result<EvaluationDetails<T>>
```

If you're primarily a Go developer, this pattern makes sense. But Rust is generally built around `Result`s, not tuples of value-error pairs. Rust has strong pattern matching functionality around its enums and the standard library's `Result` is a huge one. This isn't particularly idiomatic in that regard.

Most consumers would utilize patterns like

```rust
match client.value_details('example_flag', 'default', EvaluationContext::new()) {
  Ok(details) => ...,
  Err(err) => ...
}
```

Alternatively, the error being encapsulated with the `EvaluationDetails` would also make sense, and 'conforms' to the specification but also is less idiomatic for Rust.

Reply:
Expanding this a little further, `Result` also utilizes the `must_use` attribute which enforces any returned `Result` must be accessed (thus checked) [[Source](https://doc.rust-lang.org/std/result/#results-must-be-used)].

Reply:
I will look at reworking this - I have basically done a copy of the golang SDK and tried to make it rusty, though I've been learning rust whilst doing it haha.

Reply:
Thanks for the feedback !